File: iosApp/iosApp/Support/KoinHelper.swift
import Foundation
import Shared

/// Swift-side convenience facade over the Kotlin `KoinAccess` instance.
/// All repositories registered in the shared Koin graph are exposed here as
/// computed properties so SwiftUI views can write e.g. `koin.tickets` and get
/// a strongly-typed Kotlin repository without going through `KoinPlatform`.
///
/// The global `koin` singleton (defined in `KoinAccess+Singleton.swift`) already
/// wraps `KoinHelperKt.koinAccess()`; this file simply documents and verifies
/// that every new repository is reachable.
enum KoinHelper {
    /// Returns the singleton `KoinAccess`. Identical to the global `koin`.
    static var access: KoinAccess { koin }

    // MARK: - Convenience accessors mirroring the Kotlin contract.
    static var auth: AuthRepository { koin.auth }
    static var events: EventsRepository { koin.events }
    static var eventSchedules: EventSchedulesRepository { koin.eventSchedules }
    static var deals: DealsRepository { koin.deals }
    static var sigs: SigsRepository { koin.sigs }
    static var stamps: StampsRepository { koin.stamps }
    static var notifications: NotificationsRepository { koin.notifications }
    static var locations: LocationsRepository { koin.locations }
    static var addons: AddonsRepository { koin.addons }
    static var search: SearchRepository { koin.search }
    static var regSoci: RegSociRepository { koin.regSoci }
    static var boutique: BoutiqueRepository { koin.boutique }
    static var quid: QuidRepository { koin.quid }
    static var mensaTest: MensaTestClient { koin.mensaTest }
    static var podcasts: PodcastsRepository { koin.podcasts }
    static var documents: DocumentsRepository { koin.documents }
    static var tickets: TicketsRepository { koin.tickets }
    static var receipts: ReceiptsRepository { koin.receipts }
    static var devices: DevicesRepository { koin.devices }
    static var calendarLinks: CalendarLinksRepository { koin.calendarLinks }
    static var paymentMethods: PaymentMethodsRepository { koin.paymentMethods }
    static var metadata: MetadataRepository { koin.metadata }
    static var i18n: I18n { koin.i18n }
    static var onboarding: OnboardingState { koin.onboarding }
}

File: shared/src/commonMain/kotlin/it/mensa/shared/KoinHelper.kt
package it.mensa.shared

import it.mensa.shared.auth.AuthRepository
import it.mensa.shared.i18n.I18n
import it.mensa.shared.onboarding.OnboardingState
import it.mensa.shared.repository.AddonsRepository
import it.mensa.shared.repository.BoutiqueRepository
import it.mensa.shared.repository.CalendarLinksRepository
import it.mensa.shared.repository.DealsRepository
import it.mensa.shared.repository.DevicesRepository
import it.mensa.shared.repository.DocumentsRepository
import it.mensa.shared.repository.ExAppsRepository
import it.mensa.shared.repository.EventSchedulesRepository
import it.mensa.shared.repository.EventsRepository
import it.mensa.shared.repository.LocationsRepository
import it.mensa.shared.repository.MetadataRepository
import it.mensa.shared.repository.OrgChartRepository
import it.mensa.shared.repository.NotificationsRepository
import it.mensa.shared.repository.PaymentMethodsRepository
import it.mensa.shared.repository.ReceiptsRepository
import it.mensa.shared.repository.RegSociRepository
import it.mensa.shared.repository.SearchRepository
import it.mensa.shared.repository.SigsRepository
import it.mensa.shared.repository.StampsRepository
import it.mensa.shared.repository.LocalOfficesRepository
import it.mensa.shared.repository.PodcastsRepository
import it.mensa.shared.iqtest.MensaTestClient
import it.mensa.shared.repository.QuidRepository
import it.mensa.shared.repository.TicketsRepository
import it.mensa.shared.spotlight.SpotlightSyncEngine
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import org.koin.mp.KoinPlatform

class KoinAccess(
    val auth: AuthRepository,
    val events: EventsRepository,
    val eventSchedules: EventSchedulesRepository,
    val deals: DealsRepository,
    val sigs: SigsRepository,
    val stamps: StampsRepository,
    val notifications: NotificationsRepository,
    val locations: LocationsRepository,
    val addons: AddonsRepository,
    val search: SearchRepository,
    val regSoci: RegSociRepository,
    val boutique: BoutiqueRepository,
    val documents: DocumentsRepository,
    val tickets: TicketsRepository,
    val receipts: ReceiptsRepository,
    val devices: DevicesRepository,
    val calendarLinks: CalendarLinksRepository,
    val paymentMethods: PaymentMethodsRepository,
    val metadata: MetadataRepository,
    val orgChart: OrgChartRepository,
    val quid: QuidRepository,
    val podcasts: PodcastsRepository,
    val localOffices: LocalOfficesRepository,
    val exApps: ExAppsRepository,
    val i18n: I18n,
    val onboarding: OnboardingState,
    val mensaTest: MensaTestClient,
    val spotlightSync: SpotlightSyncEngine,
)

fun koinAccess(): KoinAccess = KoinAccess(
    auth = KoinPlatform.getKoin().get(),
    events = KoinPlatform.getKoin().get(),
    eventSchedules = KoinPlatform.getKoin().get(),
    deals = KoinPlatform.getKoin().get(),
    sigs = KoinPlatform.getKoin().get(),
    stamps = KoinPlatform.getKoin().get(),
    notifications = KoinPlatform.getKoin().get(),
    locations = KoinPlatform.getKoin().get(),
    addons = KoinPlatform.getKoin().get(),
    search = KoinPlatform.getKoin().get(),
    regSoci = KoinPlatform.getKoin().get(),
    boutique = KoinPlatform.getKoin().get(),
    documents = KoinPlatform.getKoin().get(),
    tickets = KoinPlatform.getKoin().get(),
    receipts = KoinPlatform.getKoin().get(),
    devices = KoinPlatform.getKoin().get(),
    calendarLinks = KoinPlatform.getKoin().get(),
    paymentMethods = KoinPlatform.getKoin().get(),
    metadata = KoinPlatform.getKoin().get(),
    orgChart = KoinPlatform.getKoin().get(),
    quid = KoinPlatform.getKoin().get(),
    podcasts = KoinPlatform.getKoin().get(),
    localOffices = KoinPlatform.getKoin().get(),
    exApps = KoinPlatform.getKoin().get(),
    i18n = KoinPlatform.getKoin().get(),
    onboarding = KoinPlatform.getKoin().get(),
    mensaTest = KoinPlatform.getKoin().get(),
    spotlightSync = KoinPlatform.getKoin().get(),
)

fun appScope(): CoroutineScope =
    CoroutineScope(SupervisorJob() + Dispatchers.Default)
